File: 600/652 - Find Duplicate Subtrees.java
import java.util.HashSet;
import java.util.ArrayList;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<TreeNode> findDuplicateSubtrees(TreeNode root) {
        List<TreeNode> ret = new ArrayList<TreeNode>();
        
        helper(root, ret, new HashSet<String>(), new HashSet<String>());
        
        return ret;
    }
    
    private String helper(TreeNode root, List<TreeNode> ret, HashSet<String> found, HashSet<String> added) {
        StringBuilder sb;
        String s;
        
        if (root == null)
            return "n#";
        
        sb = new StringBuilder();
        
        sb.append(root.val);
        sb.append('#');
        sb.append(helper(root.left, ret, found, added));
        sb.append(helper(root.right, ret, found, added));
        
        s = sb.toString();
        
        if (found.contains(s)) {
            if (! added.contains(s)) {
                ret.add(root);
                added.add(s);
            }
        } else
            found.add(s);
        
        return s;
    }
}

File: csharp/652.find-duplicate-subtrees.cs
/*
 * @lc app=leetcode id=652 lang=csharp
 *
 * [652] Find Duplicate Subtrees
 */

using System.Collections.Generic;

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {

    private static string FindDuplicateSubtreesRecurse(TreeNode root, IList<TreeNode> ret, IDictionary<string, int> map) {
        if (root == null) {
            return "-";
        }
        string s = $"{root.val},{FindDuplicateSubtreesRecurse(root.left, ret, map)},{FindDuplicateSubtreesRecurse(root.right, ret, map)}";
        int val = 0;
        map.TryGetValue(s, out val);
        if (val == 1) {
            ret.Add(root);
        }
        map[s] = val + 1;
        return s;
    }

    public IList<TreeNode> FindDuplicateSubtrees(TreeNode root) {
        IList<TreeNode> ret = new List<TreeNode>();
        FindDuplicateSubtreesRecurse(root, ret, new Dictionary<string, int>());
        return ret;
    }
}
// @lc code=end

